### ISSUE TITLE
plugin causes open() signature to change

### ISSUE THREAD

**Issue by arnimarj** [Shoobx/mypy-zope#16]
Hi,

I ran across an interesting issue. The signature of `open()` unexpectedly became `typing.IO[Any]` instead of `typing.TextIO` when the plugin is enabled.

`example.py`:
```python
reveal_type(open('str'))
```

`plugin.ini`:
```
[mypy]
plugins = mypy_zope:plugin
```

Here is small test:

```shell
$ more example.py 
reveal_type(open('str'))
$ more plugin.ini 
[mypy]
plugins = mypy_zope:plugin
$ mypy --config plugin.ini example.py 
example.py:1: note: Revealed type is 'typing.IO[Any]'
$ mypy example.py 
example.py:1: note: Revealed type is 'typing.TextIO'
```

I created a test-case here:

https://github.com/Shoobx/mypy-zope/compare/master...arnimarj:bug/open-signature-changes

**Comment by kedder** [comment]
Interesting. This is triggered by overriding `get_function_hook` function in a plugin. `typing.IO[Any]` is what is being passed to it the hook in `function_ctx.default_return_type`. If hook returns this default type, mypy takes it as granted. If hook is not invoked, mypy infers type as `typing.TextIO` somehow.

This looks like a mypy bug to me. I'll submit the report upstream.

**Comment by kedder** [comment]
Apparently mypy doesn't consider this as mypy issue. This bug affects a feature that allows to guess type of a field in a schema, such as 

```python
    textline = zope.schema.TextLine(
        title='Title',
        description='Optional text line',
        required=False)
```
@arnimarj is there any practical issues with the type mismatch? Do you depend on this feature?

**Comment by arnimarj** [comment]
I ran into the issue when passing down the return value of open down to a function which requires BinaryIO. This is the first time encountering this after having annotated in the order on 10k lines, so maybe not a big deal. 

So in practice not a big deal, but if we start using more plugins that may change.

**Comment by kedder** [comment]
I see. I'm going to try suggestion from https://github.com/python/mypy/issues/8331, maybe we can do something about it.

**Comment by arnimarj** [comment]
There is a open PR (https://github.com/python/mypy/pull/9275) which removes the `open` plugin. Just tried the test-case above on that branch, and got this:

```shell
$ mypy --config plugin.ini example.py 
example.py:1: note: Revealed type is 'io.TextIOWrapper'

$ mypy example.py 
example.py:1: note: Revealed type is 'typing.TextIO'
```

**Comment by arnimarj** [comment]
I just tested this on the latest mypy (0.800), and this is no longer an issue.
